File: 131.palindrome-partitioning.py
#
# @lc app=leetcode id=131 lang=python3
#
# [131] Palindrome Partitioning
#

# @lc code=start
from typing import List


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        self.dfs(s,[],res)
        return res
        
    def isPalindrome(self, s):
        return s == s[::-1]

    def dfs(self, s, path, res):
        if not s:
            res.append(path)

        for i in range(1,len(s)+1):
            if self.isPalindrome(s[:i]):
                self.dfs(s[i:], path + [s[:i]], res)
        
# @lc code=end

File: 131.palindrome-partitioning.cpp
/*
 * @lc app=leetcode id=131 lang=cpp
 *
 * [131] Palindrome Partitioning
 */

// @lc code=start
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> path;
        dfs(s, path, res);
        return res;
    }
    
private:
    bool isPalindrome(const string& s) {
        int left = 0;
        int right = s.size() - 1;
        while (left < right) {
            if (s[left] != s[right]) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    void dfs(const string& s, vector<string>& path, vector<vector<string>>& res) {
        if (s.empty()) {
            res.push_back(path);
            return;
        }
        
        for (int i = 1; i <= s.size(); i++) {
            string prefix = s.substr(0, i);
            if (isPalindrome(prefix)) {
                path.push_back(prefix);
                dfs(s.substr(i), path, res);
                path.pop_back();
            }
        }
    }
};

// @lc code=end

